pulp_ansible/app/tasks/collections.py

```diff
@@ -63,11 +63,83 @@
     parse_collections_requirements_file,
     RequirementsFileEntry,
 )
+from asgiref.sync import sync_to_async
+
+from django.db.utils import IntegrityError
+from galaxy_importer.collection import sync_collection
+from git import GitCommandError, Repo
+
+
+from uuid import uuid4
 
 
 log = logging.getLogger(__name__)
 
 
+async def declarative_content_from_git_repo(remote, url, git_ref=None, metadata_only=False):
+    """Returns a DeclarativeContent for the Collection in a Git repository."""
+    if git_ref:
+        try:
+            gitrepo = Repo.clone_from(url, uuid4(), depth=1, branch=git_ref)
+        except GitCommandError:
+            gitrepo = Repo.clone_from(url, uuid4())
+            gitrepo.git.checkout(git_ref)
+    else:
+        gitrepo = Repo.clone_from(url, uuid4(), depth=1)
+    commit_sha = gitrepo.head.commit.hexsha
+    metadata, artifact_path = sync_collection(gitrepo.working_dir, ".")
+    if not metadata_only:
+        artifact = Artifact.init_and_validate(artifact_path)
+        try:
+            await sync_to_async(artifact.save)()
+        except IntegrityError:
+            artifact = Artifact.objects.get(sha256=artifact.sha256)
+        metadata["artifact_url"] = reverse("artifacts-detail", args=[artifact.pk])
+        metadata["artifact"] = artifact
+    else:
+        metadata["artifact"] = None
+        metadata["artifact_url"] = None
+    metadata["remote_artifact_url"] = "{}/commit/{}".format(url.rstrip("/"), commit_sha)
+
+    artifact = metadata["artifact"]
+    try:
+        collection_version = await sync_to_async(create_collection_from_importer)(
+            metadata, metadata_only=metadata_only
+        )
+        await sync_to_async(ContentArtifact.objects.get_or_create)(
+            artifact=artifact,
+            content=collection_version,
+            relative_path=collection_version.relative_path,
+        )
+    except ValidationError as e:
+        if e.args[0]["non_field_errors"][0].code == "unique":
+            namespace = metadata["metadata"]["namespace"]
+            name = metadata["metadata"]["name"]
+            version = metadata["metadata"]["version"]
+        else:
+            raise e
+        collection_version = await sync_to_async(CollectionVersion.objects.get)(
+            namespace=namespace, name=name, version=version
+        )
+    if artifact is None:
+        artifact = Artifact()
+    d_artifact = DeclarativeArtifact(
+        artifact=artifact,
+        url=metadata["remote_artifact_url"],
+        relative_path=collection_version.relative_path,
+        remote=remote,
+        deferred_download=metadata_only,
+    )
+
+    # TODO: docs blob support??
```

Comment:
sync machinery will handle docs blob by trying to get data from docs blob endpoint